File: tools/pcsx-smoke.lua
-- Non-interactive PCSX-Redux probe used by tools/pcsx-smoke.py.
--
-- Addresses and limits deliberately come from the wrapper rather than being
-- copied from the retail layout: a normal relink is expected to move them.

local function requiredNumber(name)
    local value = os.getenv(name)
    local number = value and tonumber(value)
    if number == nil then
        printError("TENCHU_SMOKE FAIL missing numeric environment variable " .. name)
        PCSX.quit(2)
        error("missing " .. name)
    end
    return number
end

local function optionalNumber(name)
    local value = os.getenv(name)
    return value and tonumber(value) or nil
end

local entryAddress = requiredNumber("TENCHU_SMOKE_ENTRY")
local mainAddress = requiredNumber("TENCHU_SMOKE_MAIN")
local loopAddress = requiredNumber("TENCHU_SMOKE_LOOP")
local entryWord = optionalNumber("TENCHU_SMOKE_ENTRY_WORD")
local mainWord = optionalNumber("TENCHU_SMOKE_MAIN_WORD")
local loopWord = optionalNumber("TENCHU_SMOKE_LOOP_WORD")
local requiredFrames = requiredNumber("TENCHU_SMOKE_FRAMES")
local requiredLoops = requiredNumber("TENCHU_SMOKE_LOOPS")
local autoPad = os.getenv("TENCHU_SMOKE_AUTOPAD") == "1"
local watchCounterAddress = optionalNumber("TENCHU_SMOKE_WATCH_COUNTER")
local watchEqualsAddress = optionalNumber("TENCHU_SMOKE_WATCH_EQUALS_ADDR")
local watchEqualsValue = optionalNumber("TENCHU_SMOKE_WATCH_EQUALS_VALUE")

local ffi = require("ffi")
local function readU32(address)
    -- 2 MiB main RAM, mirrored across KUSEG/KSEG0/KSEG1 bases.
    local base = ffi.cast("uint8_t*", PCSX.getMemPtr())
    return tonumber(ffi.cast("uint32_t*", base + bit.band(address, 0x1fffff))[0])
end

local pad
local padButtons
local autoPadButtons
if autoPad then
    pad = PCSX.SIO0.slots[1].pads[1]
    padButtons = PCSX.CONSTS.PAD.BUTTON
    autoPadButtons = {padButtons.START, padButtons.CIRCLE, padButtons.CROSS}
end

local function releaseAutoPad()
    if not autoPad then return end
    for _, button in ipairs(autoPadButtons) do pad.clearOverride(button) end
end

TenchuSmoke = {
    entryHits = 0,
    mainHits = 0,
    loopHits = 0,
    totalVsyncs = 0,
    frames = 0,
    failed = false,
    passed = false,
    watchFirst = nil,
    watchLast = nil,
    equalsLast = nil,
}

local function fail(message)
    if TenchuSmoke.failed then return end
    TenchuSmoke.failed = true
    local regs = PCSX.getRegisters()
    printError(string.format(
        "TENCHU_SMOKE FAIL %s pc=0x%08x cause=0x%08x frames=%d loops=%d",
        message, tonumber(regs.pc), tonumber(regs.CP0.r[13]),
        TenchuSmoke.frames, TenchuSmoke.loopHits))
    PCSX.quit(1)
end

-- On the disc path SLPS and MENU code occupy the same RAM before MAIN.EXE
-- loads, so executing a probe address is only evidence once the instruction
-- word there matches the image the wrapper selected.
local function imageWordPresent(address, word)
    return word == nil or readU32(address) == word
end

local function oneShot(counter, breakpoint, address, word)
    return function()
        if not imageWordPresent(address, word) then return true end
        TenchuSmoke[counter] = TenchuSmoke[counter] + 1
        if counter == "mainHits" then releaseAutoPad() end
        TenchuSmoke[breakpoint]:disable()
        return true
    end
end

TenchuSmoke.entryBreakpoint = PCSX.addBreakpoint(
    entryAddress, "Exec", 4, "",
    oneShot("entryHits", "entryBreakpoint", entryAddress, entryWord),
    "Tenchu smoke entry")
TenchuSmoke.mainBreakpoint = PCSX.addBreakpoint(
    mainAddress, "Exec", 4, "",
    oneShot("mainHits", "mainBreakpoint", mainAddress, mainWord),
    "Tenchu smoke main")
TenchuSmoke.loopBreakpoint = PCSX.addBreakpoint(
    loopAddress, "Exec", 4, "", function()
        if TenchuSmoke.mainHits == 0 then return true end
        if not imageWordPresent(loopAddress, loopWord) then return true end
        TenchuSmoke.loopHits = TenchuSmoke.loopHits + 1
        if TenchuSmoke.loopHits >= requiredLoops then
            TenchuSmoke.loopBreakpoint:disable()
        end
        return true
    end, "Tenchu smoke main loop")

TenchuSmoke.pauseListener = PCSX.Events.createEventListener(
    "ExecutionFlow::Pause", function(event)
        if event.exception then
            fail("first-chance CPU exception")
        elseif not TenchuSmoke.failed and not TenchuSmoke.passed then
            fail("emulator paused unexpectedly")
        end
    end)

TenchuSmoke.vsyncListener = PCSX.Events.createEventListener(
    "GPU::Vsync", function()
        TenchuSmoke.totalVsyncs = TenchuSmoke.totalVsyncs + 1
        if autoPad and TenchuSmoke.mainHits == 0 then
            -- The disc path passes through SLPS and MENU before MAIN.EXE.  A
            -- short, separated START/CIRCLE/CROSS pulse train advances their
            -- default choices without assuming the confirm-button region.
            local phase = TenchuSmoke.totalVsyncs % 90
            if phase == 1 then
                pad.setOverride(padButtons.START)
            elseif phase == 4 then
                pad.clearOverride(padButtons.START)
            elseif phase == 31 then
                pad.setOverride(padButtons.CIRCLE)
            elseif phase == 34 then
                pad.clearOverride(padButtons.CIRCLE)
            elseif phase == 61 then
                pad.setOverride(padButtons.CROSS)
            elseif phase == 64 then
                pad.clearOverride(padButtons.CROSS)
            end
        end
        -- OpenBIOS and Tenchu's one-time startup both generate VSyncs.  Count
        -- only frames after the per-frame loop probe has actually fired.
        if TenchuSmoke.loopHits == 0 then return end
        TenchuSmoke.frames = TenchuSmoke.frames + 1
        if watchCounterAddress then
            local sample = readU32(watchCounterAddress)
            if TenchuSmoke.watchFirst == nil then
                TenchuSmoke.watchFirst = sample
            end
            TenchuSmoke.watchLast = sample
        end
        if watchEqualsAddress then
            TenchuSmoke.equalsLast = readU32(watchEqualsAddress)
        end
        if TenchuSmoke.frames < requiredFrames then return end
        local watchCounterOk = not watchCounterAddress or
            (TenchuSmoke.watchLast ~= nil and TenchuSmoke.watchLast > 0 and
             TenchuSmoke.watchLast > TenchuSmoke.watchFirst)
        local watchEqualsOk = not watchEqualsAddress or
            TenchuSmoke.equalsLast == watchEqualsValue
        if TenchuSmoke.entryHits ~= 0 and
           TenchuSmoke.mainHits ~= 0 and
           TenchuSmoke.loopHits >= requiredLoops and
           watchCounterOk and watchEqualsOk then
            TenchuSmoke.passed = true
            local watchReport = ""
            if watchCounterAddress then
                watchReport = string.format(
                    " watchCounter=%d..%d",
                    TenchuSmoke.watchFirst, TenchuSmoke.watchLast)
            end
            if watchEqualsAddress then
                watchReport = watchReport ..
                    string.format(" watchEquals=0x%08x", TenchuSmoke.equalsLast)
            end
            print(string.format(
                "TENCHU_SMOKE PASS entry=%d main=%d frames=%d loops=%d cycles=%s%s",
                TenchuSmoke.entryHits, TenchuSmoke.mainHits, TenchuSmoke.frames,
                TenchuSmoke.loopHits, tostring(PCSX.getCPUCycles()), watchReport))
            PCSX.quit(0)
        elseif TenchuSmoke.frames > requiredFrames + 600 then
            -- The base conditions or a memory watch never converged within a
            -- generous post-loop window; report what was actually observed.
            fail(string.format(
                "watch conditions unmet counter=%s..%s equals=%s",
                tostring(TenchuSmoke.watchFirst), tostring(TenchuSmoke.watchLast),
                tostring(TenchuSmoke.equalsLast)))
        end
    end)

local armedWatch = ""
if watchCounterAddress then
    armedWatch = string.format(" watchCounter=0x%08x", watchCounterAddress)
end
if watchEqualsAddress then
    armedWatch = armedWatch .. string.format(
        " watchEquals=0x%08x==0x%08x", watchEqualsAddress, watchEqualsValue)
end
print(string.format(
    "TENCHU_SMOKE armed entry=0x%08x main=0x%08x loop=0x%08x frames=%d loops=%d%s",
    entryAddress, mainAddress, loopAddress, requiredFrames, requiredLoops,
    armedWatch))

File: tools/pcsx_smoke.py
#!/usr/bin/env python3
"""Run a bounded, non-interactive PCSX-Redux smoke test of ``main_relink.exe``.

By default the emulator loads the original disc for game assets but injects
the supplied PS-X EXE at the BIOS shell, matching the fast ``./Build run``
path.  ``--repack`` instead builds a temporary disc and exercises the full
``SLPS -> MENU -> MAIN`` handoff.  A Lua probe derives all code addresses from
the current executable/ELF pair and requires the executable entry, ``main``,
repeated ``PadProc`` calls from the main loop, post-loop VSyncs, and no
first-chance CPU exception.

This is deliberately a runtime sanity gate, not a byte-match check or proof of
the complete launcher/movie/audio path.  It never patches the executable.
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
import os
from pathlib import Path
import shutil
import signal
import struct
import subprocess
import sys
import tempfile
from typing import Mapping, Sequence

try:
    from tools import psxexe
except ImportError:  # Direct execution (``tools/pcsx_smoke.py``).
    import psxexe  # type: ignore[no-redef]


ROOT = Path(__file__).resolve().parent.parent
DEFAULT_EXE = ROOT / ".shake" / "build" / "tenchu" / "main_relink.exe"
DEFAULT_ELF = ROOT / ".shake" / "build" / "tenchu" / "main_relink.exe.elf"
LUA_PROBE = Path(__file__).with_name("pcsx-smoke.lua")
PSX_EXE_MAGIC = b"PS-X EXE"
PSX_EXE_ENTRY_OFFSET = 0x10


class SmokeError(RuntimeError):
    """An invalid input or failed host-side smoke-test invariant."""


@dataclass(frozen=True)
class ProbeAddresses:
    entry: int
    main: int
    loop: int
    watch_counter: int | None = None
    watch_equals: int | None = None
    watch_equals_value: int | None = None


@dataclass(frozen=True)
class ProcessResult:
    returncode: int
    output: str
    timed_out: bool = False


def _existing_file(path: str | Path, description: str) -> Path:
    candidate = Path(path).expanduser().resolve()
    if not candidate.is_file():
        raise SmokeError(f"{description} not found: {candidate}")
    return candidate


def _existing_executable(path: str | Path, description: str) -> Path:
    candidate = _existing_file(path, description)
    if not os.access(candidate, os.X_OK):
        raise SmokeError(f"{description} is not executable: {candidate}")
    return candidate


def discover_pcsx(
    explicit: str | Path | None,
    *,
    environ: Mapping[str, str] = os.environ,
    home: Path | None = None,
) -> Path:
    """Use the same explicit/env/PATH/checkout preference as ``./Build run``."""
    if explicit is not None:
        return _existing_executable(explicit, "PCSX-Redux executable")
    configured = environ.get("PCSX_REDUX")
    if configured:
        return _existing_executable(configured, "$PCSX_REDUX")
    on_path = shutil.which("pcsx-redux")
    if on_path:
        return _existing_executable(on_path, "pcsx-redux on PATH")
    fallback = (home or Path.home()) / "programming" / "pcsx-redux" / "pcsx-redux"
    if fallback.is_file() and os.access(fallback, os.X_OK):
        return fallback.resolve()
    raise SmokeError(
        "pcsx-redux not found; pass --pcsx, set PCSX_REDUX, put it on PATH, "
        "or build it under ~/programming/pcsx-redux"
    )


def discover_cue(
    explicit: str | Path | None,
    *,
    environ: Mapping[str, str] = os.environ,
    root: Path = ROOT,
    home: Path | None = None,
) -> Path:
    """Find the user-supplied original disc using the documented search order."""
    if explicit is not None:
        return _existing_file(explicit, "original disc cue")
    configured = environ.get("TENCHU_CUE")
    if configured:
        return _existing_file(configured, "$TENCHU_CUE")
    for directory in (root / "disks", (home or Path.home()) / "tenchu-iso"):
        if not directory.is_dir():
            continue
        cues = sorted(
            path for path in directory.iterdir()
            if path.is_file() and path.suffix.lower() == ".cue"
        )
        if cues:
            return cues[0].resolve()
    raise SmokeError(
        "original disc .cue not found; pass --cue, set TENCHU_CUE, or place "
        "it under disks/ or ~/tenchu-iso/"
    )


def read_psx_exe_entry(path: Path) -> int:
    with path.open("rb") as source:
        header = source.read(PSX_EXE_ENTRY_OFFSET + 4)
    if len(header) < PSX_EXE_ENTRY_OFFSET + 4 or header[:8] != PSX_EXE_MAGIC:
        raise SmokeError(f"{path}: not a complete PS-X EXE header")
    entry = struct.unpack_from("<I", header, PSX_EXE_ENTRY_OFFSET)[0]
    if entry == 0:
        raise SmokeError(f"{path}: PS-X EXE entry is zero")
    return entry


def read_exe_word(path: Path, address: int) -> int:
    """Read the u32 the PS-X EXE will place at ``address`` once loaded.

    The probe breakpoints verify these image words at runtime: on the disc
    path SLPS/MENU code occupies the same RAM before MAIN.EXE, so an
    execution of the bare address is not yet evidence that our executable
    is running.
    """
    data = path.read_bytes()
    if len(data) < 0x800 or data[:8] != PSX_EXE_MAGIC:
        raise SmokeError(f"{path}: not a complete PS-X EXE header")
    t_addr, t_size = struct.unpack_from("<II", data, 0x18)
    if not t_addr <= address <= t_addr + t_size - 4:
        raise SmokeError(
            f"{path}: address 0x{address:08x} is outside the loaded image "
            f"0x{t_addr:08x}..0x{t_addr + t_size:08x}"
        )
    offset = 0x800 + (address - t_addr)
    if offset + 4 > len(data):
        raise SmokeError(f"{path}: image truncated at 0x{address:08x}")
    return struct.unpack_from("<I", data, offset)[0]


def parse_watch_equals(spec: str) -> tuple[str, int]:
    """Parse ``SYMBOL=VALUE`` (any int base) for the equality watch."""
    symbol, separator, raw_value = spec.partition("=")
    if not separator or not symbol or not raw_value:
        raise SmokeError(f"--watch-equals expects SYMBOL=VALUE, got: {spec}")
    try:
        value = int(raw_value, 0)
    except ValueError as error:
        raise SmokeError(f"--watch-equals value is not an integer: {spec}") from error
    if not 0 <= value <= 0xFFFFFFFF:
        raise SmokeError(f"--watch-equals value must fit in 32 bits: {spec}")
    return symbol, value


def resolve_probe_addresses(
    exe: Path,
    elf: Path,
    *,
    main_symbol: str = "main",
    loop_symbol: str = "PadProc",
    watch_counter_symbol: str | None = None,
    watch_equals: tuple[str, int] | None = None,
) -> ProbeAddresses:
    entry = read_psx_exe_entry(exe)
    try:
        metadata = psxexe.read_elf_metadata(elf)
        main = metadata.symbol(main_symbol)
        loop = metadata.symbol(loop_symbol)
        watch_counter = (
            metadata.symbol(watch_counter_symbol)
            if watch_counter_symbol is not None
            else None
        )
        watch_equals_address = (
            metadata.symbol(watch_equals[0]) if watch_equals is not None else None
        )
    except (OSError, psxexe.PsxExeError) as error:
        raise SmokeError(str(error)) from error
    return ProbeAddresses(
        entry=entry,
        main=main,
        loop=loop,
        watch_counter=watch_counter,
        watch_equals=watch_equals_address,
        watch_equals_value=watch_equals[1] if watch_equals is not None else None,
    )


def build_command(
    pcsx: Path,
    cue: Path,
    exe: Path,
    temp: Path,
    *,
    bios: Path | None = None,
    load_exe: bool = True,
    lua_probe: Path | None = None,
) -> list[str]:
    command = [
        str(pcsx),
        "-safe",
        "-no-ui",
        "-noupdate",
        "-no-fastboot",
        "-2mb",
        "-run",
        "-iso",
        str(cue),
        # Lua execution breakpoints require the debugger-enabled interpreter.
        # Do not use -testmode: it suppresses first-chance exception pauses,
        # which are one of the failure signals the Lua probe observes.
        "-interpreter",
        "-debugger",
        "-softgpu",
        "-memcard1",
        str(temp / "memcard1.mcd"),
        "-memcard2",
        str(temp / "memcard2.mcd"),
    ]
    if load_exe:
        command.extend(("-loadexe", str(exe)))
    if bios is not None:
        command.extend(("-bios", str(bios)))
    command.extend(("-dofile", str((lua_probe or LUA_PROBE).resolve())))
    return command


def package_disc(
    exe: Path,
    original_cue: Path,
    temp: Path,
    *,
    environ: Mapping[str, str],
    timeout: float,
) -> tuple[Path, ProcessResult]:
    """Build a temporary disc with mkiso.py, including its grown auto-LBA path."""
    if shutil.which("mkpsxiso") is None:
        raise SmokeError("--repack needs mkpsxiso; run the smoke test inside `nix develop`")
    output_prefix = temp / "tenchu-smoke"
    command = [
        sys.executable,
        str(ROOT / "tools" / "mkiso.py"),
        "--exe",
        str(exe),
        "-o",
        str(output_prefix),
    ]
    package_environment = dict(environ)
    package_environment["TENCHU_CUE"] = str(original_cue)
    result = run_process(
        command, cwd=ROOT, environ=package_environment, timeout=timeout
    )
    return output_prefix.with_suffix(".cue"), result


def _terminate_process_group(process: subprocess.Popen[str]) -> None:
    if process.poll() is not None:
        return
    if os.name == "posix":
        try:
            os.killpg(process.pid, signal.SIGTERM)
        except ProcessLookupError:
            pass
    else:  # pragma: no cover - the project runtime lane is Linux-first.
        process.terminate()


def _kill_process_group(process: subprocess.Popen[str]) -> None:
    if process.poll() is not None:
        return
    if os.name == "posix":
        try:
            os.killpg(process.pid, signal.SIGKILL)
        except ProcessLookupError:
            pass
    else:  # pragma: no cover - the project runtime lane is Linux-first.
        process.kill()


def run_process(
    command: Sequence[str],
    *,
    cwd: Path,
    environ: Mapping[str, str],
    timeout: float,
) -> ProcessResult:
    try:
        process = subprocess.Popen(
            list(command),
            cwd=cwd,
            env=dict(environ),
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            start_new_session=True,
        )
    except OSError as error:
        raise SmokeError(f"could not start {command[0]}: {error}") from error
    try:
        output, _ = process.communicate(timeout=timeout)
        return ProcessResult(process.returncode, output)
    except subprocess.TimeoutExpired:
        _terminate_process_group(process)
        try:
            output, _ = process.communicate(timeout=5)
        except subprocess.TimeoutExpired:
            _kill_process_group(process)
            output, _ = process.communicate()
        return ProcessResult(process.returncode, output, timed_out=True)


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--exe", type=Path, default=DEFAULT_EXE)
    parser.add_argument("--elf", type=Path, default=DEFAULT_ELF)
    parser.add_argument("--cue", type=Path, help="original disc .cue")
    parser.add_argument("--pcsx", type=Path, help="PCSX-Redux executable")
    parser.add_argument("--bios", type=Path, help="optional real PS1 BIOS")
    parser.add_argument(
        "--repack",
        action="store_true",
        help=(
            "temporarily repack the selected EXE and boot the full "
            "SLPS->MENU->MAIN disc path"
        ),
    )
    parser.add_argument(
        "--timeout", type=float,
        help="emulator timeout (default: 60s direct, 180s with --repack)",
    )
    parser.add_argument("--repack-timeout", type=float, default=300.0)
    parser.add_argument("--frames", type=int, default=12,
                        help="VSyncs required after reaching the main loop")
    parser.add_argument("--loop-hits", type=int, default=3,
                        help="required calls to the main-loop PadProc probe")
    parser.add_argument("--main-symbol", default="main")
    parser.add_argument("--loop-symbol", default="PadProc")
    parser.add_argument(
        "--watch-counter", metavar="SYMBOL",
        help=(
            "require the u32 at SYMBOL (resolved from the selected ELF) to be "
            "nonzero and to increase while the main loop runs"
        ),
    )
    parser.add_argument(
        "--watch-equals", metavar="SYMBOL=VALUE",
        help=(
            "require the u32 at SYMBOL (resolved from the selected ELF) to "
            "equal VALUE once the game is running"
        ),
    )
    args = parser.parse_args(argv)
    if args.timeout is not None and args.timeout <= 0:
        parser.error("--timeout must be positive")
    if args.repack_timeout <= 0:
        parser.error("--repack-timeout must be positive")
    if args.frames <= 0:
        parser.error("--frames must be positive")
    if args.loop_hits <= 0:
        parser.error("--loop-hits must be positive")
    return args


def run(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        exe = _existing_file(args.exe, "relinked PS-X EXE")
        elf = _existing_file(args.elf, "relinked ELF")
        pcsx = discover_pcsx(args.pcsx)
        cue = discover_cue(args.cue)
        bios = _existing_file(args.bios, "BIOS") if args.bios else None
        _existing_file(LUA_PROBE, "PCSX smoke Lua probe")
        watch_equals = (
            parse_watch_equals(args.watch_equals)
            if args.watch_equals is not None
            else None
        )
        addresses = resolve_probe_addresses(
            exe, elf, main_symbol=args.main_symbol, loop_symbol=args.loop_symbol,
            watch_counter_symbol=args.watch_counter, watch_equals=watch_equals,
        )
    except SmokeError as error:
        print(f"pcsx-smoke: {error}", file=sys.stderr)
        return 2

    try:
        entry_word = read_exe_word(exe, addresses.entry)
        main_word = read_exe_word(exe, addresses.main)
        loop_word = read_exe_word(exe, addresses.loop)
    except SmokeError as error:
        print(f"pcsx-smoke: {error}", file=sys.stderr)
        return 2

    environment = os.environ.copy()
    environment.update(
        {
            "TENCHU_SMOKE_ENTRY": str(addresses.entry),
            "TENCHU_SMOKE_MAIN": str(addresses.main),
            "TENCHU_SMOKE_LOOP": str(addresses.loop),
            "TENCHU_SMOKE_ENTRY_WORD": str(entry_word),
            "TENCHU_SMOKE_MAIN_WORD": str(main_word),
            "TENCHU_SMOKE_LOOP_WORD": str(loop_word),
            "TENCHU_SMOKE_FRAMES": str(args.frames),
            "TENCHU_SMOKE_LOOPS": str(args.loop_hits),
            "TENCHU_SMOKE_AUTOPAD": "1" if args.repack else "0",
        }
    )
    if addresses.watch_counter is not None:
        environment["TENCHU_SMOKE_WATCH_COUNTER"] = str(addresses.watch_counter)
    if addresses.watch_equals is not None:
        environment["TENCHU_SMOKE_WATCH_EQUALS_ADDR"] = str(addresses.watch_equals)
        environment["TENCHU_SMOKE_WATCH_EQUALS_VALUE"] = str(
            addresses.watch_equals_value
        )

    emulator_timeout = (
        args.timeout if args.timeout is not None else (180.0 if args.repack else 60.0)
    )
    with tempfile.TemporaryDirectory(prefix="tenchu-pcsx-smoke-") as temporary:
        temp = Path(temporary)
        disc = cue
        if args.repack:
            try:
                disc, package_result = package_disc(
                    exe, cue, temp, environ=environment, timeout=args.repack_timeout
                )
            except SmokeError as error:
                print(f"pcsx-smoke: {error}", file=sys.stderr)
                return 2
            if package_result.output:
                print(
                    package_result.output,
                    end="" if package_result.output.endswith("\n") else "\n",
                )
            if package_result.timed_out:
                print(
                    f"pcsx-smoke: disc repack timed out after {args.repack_timeout:g} seconds",
                    file=sys.stderr,
                )
                return 124
            if package_result.returncode != 0 or not disc.is_file():
                print(
                    f"pcsx-smoke: disc repack failed with status {package_result.returncode}",
                    file=sys.stderr,
                )
                return 1

        command = build_command(
            pcsx, disc, exe, temp, bios=bios, load_exe=not args.repack
        )
        try:
            result = run_process(
                command, cwd=temp, environ=environment, timeout=emulator_timeout
            )
        except SmokeError as error:
            print(f"pcsx-smoke: {error}", file=sys.stderr)
            return 2

    if result.output:
        print(result.output, end="" if result.output.endswith("\n") else "\n")
    if result.timed_out:
        print(
            f"pcsx-smoke: timed out after {emulator_timeout:g} seconds",
            file=sys.stderr,
        )
        return 124
    if result.returncode != 0:
        print(
            f"pcsx-smoke: PCSX-Redux exited with status {result.returncode}",
            file=sys.stderr,
        )
        return 1
    if "TENCHU_SMOKE FAIL " in result.output:
        print("pcsx-smoke: Lua probe reported failure", file=sys.stderr)
        return 1
    if "TENCHU_SMOKE PASS " not in result.output:
        print("pcsx-smoke: PCSX-Redux exited without the Lua PASS marker", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
